google-cloud-tools-plugin/testSrc/com/intellij/appengine/PicoContainerTestUtil.java

```diff
@@ -0,0 +1,98 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.intellij.appengine;
+
+import com.google.auto.value.AutoValue;
+
+import com.intellij.openapi.application.ApplicationManager;
+
+import org.picocontainer.MutablePicoContainer;
+
+import java.util.ArrayList;
+
+/**
+ * Handles modifications to the {@link MutablePicoContainer} stored in the {@link
+ * ApplicationManager} for tests that wish to mock registered components.
+ */
+public final class PicoContainerTestUtil {
+
+  private static final PicoContainerTestUtil INSTANCE = new PicoContainerTestUtil();
+
+  private final ArrayList<Component> components = new ArrayList<>();
+
+  /** Prevents instantiation. */
+  private PicoContainerTestUtil() {}
+
+  /** Returns the static instance of this class. */
+  public static PicoContainerTestUtil getInstance() {
+    return INSTANCE;
+  }
+
+  /**
+   * Replaces the registered component in the {@link MutablePicoContainer} with the given mocked
+   * instance.
+   *
+   * <p>You should always call {@link #tearDown()} in the test's tear-down process if you make a
+   * call to this method.
+   *
+   * @param clazz the class of the registered component
+   * @param mockedInstance the mocked instance to register
+   * @param <T> the type of the registered component
+   */
+  public <T> void replaceComponentWithMock(Class<T> clazz, T mockedInstance) {
+    Object originalInstance = setComponent(clazz, mockedInstance);
+    components.add(Component.create(clazz, originalInstance));
+  }
+
+  /** Tears down this utility's state by re-registering all of the original component instances. */
+  public void tearDown() {
+    components.forEach(component -> setComponent(component.clazz(), component.originalInstance()));
+    components.clear();
+  }
+
+  /**
+   * Replaces the component binding in the {@link MutablePicoContainer} with the given instance and
+   * returns the original component instance.
+   *
+   * @param clazz the class of the registered component
+   * @param newInstance the new instance to register
+   */
+  private static Object setComponent(Class<?> clazz, Object newInstance) {
+    MutablePicoContainer applicationContainer =
+        (MutablePicoContainer) ApplicationManager.getApplication().getPicoContainer();
+    Object originalInstance = applicationContainer.getComponentInstanceOfType(clazz);
+    applicationContainer.unregisterComponent(clazz.getName());
+    applicationContainer.registerComponentInstance(clazz.getName(), newInstance);
+    return originalInstance;
+  }
+
+  /** Represents a bound component in the {@link MutablePicoContainer}. */
+  @AutoValue
+  abstract static class Component {
```

Comment:
I thought the same thing, but figured we could benefit from it elsewhere potentially in the future. It's also only marked as a test dep so it won't affect the plugin's built size.